x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/stream_active_record.ts

```diff
@@ -0,0 +1,155 @@
+  public get definition() {
+    return this._definition;
+  }
+
+  public applyChange(change: StreamChange, desiredState: State, startingState: State): void {
+    if (change.type === 'delete') {
+      this.delete(desiredState, startingState);
+    } else {
+      this.upsert(change.request.stream, desiredState, startingState);
+    }
+  }
+
+  public delete(desiredState: State, startingState: State) {
+    try {
+      this.doDelete(desiredState, startingState);
+      this.changeStatus = 'deleted';
+    } catch (error) {
+      // Here we might return { changedSuccessfully: boolean; errors: Error[] }
+    }
+  }
+
+  public upsert(definition: StreamDefinition, desiredState: State, startingState: State) {
+    try {
+      this.doUpsert(definition, desiredState, startingState);
+      this.changeStatus = 'upserted';
+    } catch (error) {
+      // Here we might return { changedSuccessfully: boolean; errors: Error[] }
+    }
+  }
+
+  public async validate(
+    desiredState: State,
+    startingState: State,
+    scopedClusterClient: IScopedClusterClient
+  ): Promise<ValidationResult> {
+    try {
+      return this.doValidate(desiredState, startingState, scopedClusterClient);
+    } catch (error) {
+      return { isValid: false, errors: [error.message] };
+    }
+  }
+
+  public async commit(
+    storageClient: StreamsStorageClient,
+    logger: Logger,
+    scopedClusterClient: IScopedClusterClient,
+    isServerless: boolean
+  ) {
+    try {
+      this.commitStatus = 'committing';
+      if (this.changeStatus === 'upserted') {
+        this.doCommitUpsert(storageClient, logger, scopedClusterClient, isServerless);
+      } else if (this.changeStatus === 'deleted') {
+        this.doCommitDelete(storageClient, logger, scopedClusterClient, isServerless);
+      }
+      this.commitStatus = 'committed';
+    } catch (error) {
+      // We should probably grab some error here?
+      this.commitStatus = 'failed';
+    }
+  }
+
+  public async revert(
+    desiredState: State,
+    startingState: State,
+    storageClient: StreamsStorageClient,
+    logger: Logger,
+    scopedClusterClient: IScopedClusterClient,
+    isServerless: boolean
+  ) {
+    if (this.changeStatus === 'upserted') {
+      if (startingState.has(this.definition.name)) {
+        // Stream was updated, revert to previous state
+        const startingStateDefinition = startingState.get(this.definition.name)?.definition!;
+        this.upsert(startingStateDefinition, startingState, desiredState);
+      } else {
+        // Stream was created, delete it
+        this.delete(startingState, desiredState);
+      }
+    } else if (this.changeStatus === 'deleted') {
+      // Stream was deleted, revert to previous state
+      const startingStateDefinition = startingState.get(this.definition.name)?.definition!;
+      this.upsert(startingStateDefinition, startingState, desiredState);
+    }
+
+    await this.commit(storageClient, logger, scopedClusterClient, isServerless);
```

Comment:
Maybe this is a bit too convoluted.
Maybe it's sufficient to just grab the stream versions from the starting state and force commit those.
And then have some special handling for the clean up of streams that we failed to create.